File: examples/python/advanced/error-handling/multiple-error-types.py
from aws_durable_execution_sdk_python import (
    DurableContext,
    durable_execution,
    durable_step,
    ExecutionError,
    InvocationError,
    ValidationError,
    StepContext,
)

@durable_step
def complex_operation(step_context: StepContext, data: dict) -> dict:
    """Operation with multiple failure modes."""
    # Validate input
    if not data:
        raise ValueError("Data is required")
    
    # Check business rules
    if data.get("amount", 0) < 0:
        raise ExecutionError("Amount must be positive")
    
    # Call external service
    try:
        result = call_external_service(data)
        return result
    except ConnectionError:
        # Transient failure
        raise InvocationError("Service unavailable")

@durable_execution
def handler(event: dict, context: DurableContext) -> dict:
    """Handle multiple error types."""
    try:
        result = context.step(complex_operation(event))
        return {"status": "success", "result": result}
    except ValueError as e:
        return {"status": "invalid", "error": str(e)}
    except ExecutionError as e:
        return {"status": "failed", "error": str(e)}
    except InvocationError as e:
        # Let Lambda retry
        raise

File: examples/java/advanced/error-handling/multiple-error-types.java
public Map<String, Object> handleRequest(Map<String, Object> event, DurableContext ctx) {
    try {
        var result = ctx.step("complex_operation", Map.class, stepCtx -> {
            // Validate input
            if (event == null || event.isEmpty()) {
                throw new IllegalArgumentException("Data is required");
            }

            // Check business rules
            var amount = ((Number) event.getOrDefault("amount", 0)).doubleValue();
            if (amount < 0) {
                throw new StepFailedException("Amount must be positive");
            }

            // Call external service
            return callExternalService(event);
        });
        return Map.of("status", "success", "result", result);
    } catch (IllegalArgumentException e) {
        return Map.of("status", "invalid", "error", e.getMessage());
    } catch (StepFailedException e) {
        return Map.of("status", "failed", "error", e.getMessage());
    } catch (RuntimeException e) {
        // Let Lambda retry
        throw e;
    }
}
